### Student Name & ID : ARJUN KC(8773456)

### Lab 5 - Cross-Validation for Model Selection

In [53]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn import datasets
from sklearn import preprocessing
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn import metrics
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_squared_error
from sklearn.metrics import mean_absolute_percentage_error
from sklearn.model_selection import cross_val_score
from sklearn.pipeline import make_pipeline
sns.set_style("whitegrid")

### Loading Dataset

### 1)Utilize the diabetes dataset from lab 4. Perform cross-validation on nine polynomial models, ranging from degree 0 to 8

In [54]:
X, y = datasets.load_diabetes(as_frame=True, scaled=False, return_X_y=True)

In [55]:
degrees = np.arange(0, 9)  ## Loop over degrees to test
for degree in degrees:
  model = make_pipeline(PolynomialFeatures(degree), LinearRegression())  # Create a pipeline that creates polynomial features, then applies linear regression

### 2)Construct a table summarizing the cross-validation results. Each model should have a separate row in the table. Include the R-Squared and Mean Absolute Error (MAE) metrics for each model. Calculate the mean value and standard deviation of these metrics from the cross-validation. Include both values

In [56]:
cross_val_scores = [] #creating the empty list
for degree in degrees:
    # Create a pipeline that creates polynomial features, then applies linear regression
    model = make_pipeline(PolynomialFeatures(degree), LinearRegression())
    
    # Use cross-validation to compute the R-square(R^2) across multiple splits of the data
    score = cross_val_score(model, X, y, cv=5,scoring="r2")
    r2_mean = np.mean(score) # Calculate the mean R-square across all folds
    r2_std = np.std(score) # Calculate the standar-deviation R-square across all folds
    # Use cross-validation to compute the Mean Ablsoute Error(MAE) across multiple splits of the data
    scores = -(cross_val_score(model, X, y, cv=5,scoring="neg_mean_absolute_error"))
    MEA_mean = np.mean(scores) # Calculate the mean MAE across all folds
    MEA_std = np.std(scores) #Calculate the standard deviation MAE across all folds
    cross_val_scores.append({"Degree": degree,"R_Square_Mean":r2_mean,"R_Square_STD":r2_std,"MEA_Mean":MEA_mean,"MEA_STD":MEA_std}) #appending all the metrics of R-square and MAE



### Converting the cross-validation results into a DataFrame

In [59]:
cross_val_scores_df = pd.DataFrame(cross_val_scores)
cross_val_scores_df

Degree  R_Square_Mean  R_Square_STD     MEA_Mean     MEA_STD
0       0      -0.027506      0.036772    66.045624    3.474660
1       1       0.482316      0.049269    44.276499    2.100110
2       2      -0.347201      0.646338    63.546355   14.945520
3       3    -203.419848    225.878682   342.052617  142.438549
4       4    -571.083108    369.891883   657.260477  159.475902
5       5    -436.856887    379.100423   562.993636   59.917202
6       6   -1695.477360   2632.416214   742.560741  191.043026
7       7   -5530.894074   9518.586953  1032.681725  393.439617
8       8  -16076.255118  28049.952581  1475.658532  706.280274

### 3)Identification of the Best Model: Identify the model that exhibits the highest performance based on the R-Squared and MAE metrics. Provide an explanation for choosing this specific model

In [60]:
Best_Degree_R_Square = cross_val_scores_df["Degree"].iloc[1]
print("BEST Performance for R_Square Metrics among polynomial Degree 0-9  is :",Best_Degree_R_Square)
R_mean = cross_val_scores_df["R_Square_Mean"].iloc[1]
print("The R_Square metric of Mean is:",round(R_mean,2))
R_STD = cross_val_scores_df["R_Square_STD"].iloc[1]
print("The R_Square metric of Standard-deviation is",round(R_STD,3))
print("\n")
Best_Degree_MAE = cross_val_scores_df["Degree"].iloc[1]
print("BEST Performance for MAE Metrics among polynomial Degree 0-9  is :",Best_Degree_MAE)
MEA_MEAN = cross_val_scores_df["MEA_Mean"].iloc[1]
print("The MEA metric of Mean is:",round(MEA_MEAN,2))
MEA_STD = cross_val_scores_df["MEA_STD"].iloc[1]
print("The MEA metric of Standard-Deviation is:",round(MEA_STD,3))

BEST Performance for R_Square Metrics among polynomial Degree 0-9  is : 1
The R_Square metric of Mean is: 0.48
The R_Square metric of Standard-deviation is 0.049


BEST Performance for MAE Metrics among polynomial Degree 0-9  is : 1
The MEA metric of Mean is: 44.28
The MEA metric of Standard-Deviation is: 2.1


#### Among R_Squre and MAE, I would choose for R_Square metrics of Mean as it has  value of 0.48  means that approximately 48% of the variance in the dependent variable can be explained by the independent variables.
#### Also, R-squared ranges from 0 to 1, with higher values indicating a better fit for regression model which mean our data is better fited for this model.

#### Actually,the choice between R-squared and MAE depends on our specific use case and what we value more: for the better fit we use (R-squared) and for the accuracy of predictions we use (MAE).